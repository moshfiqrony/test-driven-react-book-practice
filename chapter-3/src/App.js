import React, { useState } from 'react';
import Button from './components/Button';
import Slide from './components/Slide';

function App(props) {
    const [current, setCurrent] = useState(0)

    const images = [
        {
            url: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQYmn-AewAIxk9ATJC3VOb26sVuSDkdqzDNieX9pov466JFno1k&usqp=CAU",
            description: "AI Image Enlarger - Image Enlarger Online Tool Free"
        },
        {
            url: "https://image.shutterstock.com/image-photo/beautiful-pink-flower-anemones-fresh-260nw-1028135845.jpg",
            description: "Seeing the unseeable: that 'black hole image' explained - The ..."
        },
        {
            url: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSJU95RxhFxeMd3gCBva9O6b2-ByPOwe9q__TweIor_uLtrB6jr&usqp=CAU",
            description: "Color palette generator | Canva Colors"
        },

    ]

    

    return (
        <div>
            <Slide imgUrl={images[current].url} description={images[current].description}/>
            <Button data-testid='previous' onClick={() => setCurrent(prevCurrent => prevCurrent === 0 ? images.length-1 : prevCurrent-1 )} type='button'>Previous</Button>
            <Button data-testid='next' onClick={() => setCurrent(prevCurrent => prevCurrent === images.length-1 ? 0 : prevCurrent+1 )} type='button'>Next</Button>
        </div>
    );
}

export default App;