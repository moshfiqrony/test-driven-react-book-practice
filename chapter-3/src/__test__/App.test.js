import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import App from '../App';

const images = [
    {
        url: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQYmn-AewAIxk9ATJC3VOb26sVuSDkdqzDNieX9pov466JFno1k&usqp=CAU",
        description: "AI Image Enlarger - Image Enlarger Online Tool Free"
    },
    {
        url: "https://image.shutterstock.com/image-photo/beautiful-pink-flower-anemones-fresh-260nw-1028135845.jpg",
        description: "Seeing the unseeable: that 'black hole image' explained - The ..."
    },
    {
        url: "https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSJU95RxhFxeMd3gCBva9O6b2-ByPOwe9q__TweIor_uLtrB6jr&usqp=CAU",
        description: "Color palette generator | Canva Colors"
    },

]

describe("App's architecture test", () => {
    test('is the previous button are rendering properly', () => {
        const {getByText} = render(<App/>)
        expect(getByText('Previous').type).toBe('button')
    })

    test('is the next button are rendering properly', () => {
        const {getByText} = render(<App/>)
        expect(getByText('Next').type).toBe('button')
    })

    test('are the Figure -> image and figcaption rendered correctly?', () => {
        const {container} = render(<App/>)
        expect(container.querySelector('figure')).toMatchSnapshot();
    })
})

describe("initial image description test rendered properly", () => {
    test('is the image has url props', () => {
        const {getByTestId} = render(<App/>)
        expect(getByTestId('image-test').src).toBe(images[0].url)
    })

    test('is the image has description props', () => {
        const {getByTestId} = render(<App/>)
        expect(getByTestId('image-test').alt).toBe(images[0].description)
    })
})

describe("click event for next and prev button test", () => {
    test('next button click event test', () => {
        const {getByTestId} = render(<App/>)
        const button = getByTestId('next')
        expect(getByTestId('image-test').src).toBe(images[0].url)
        expect(getByTestId('image-test').alt).toBe(images[0].description)
        fireEvent.click(button)
        expect(getByTestId('image-test').src).toBe(images[1].url)
        expect(getByTestId('image-test').alt).toBe(images[1].description)
        fireEvent.click(button)
        expect(getByTestId('image-test').src).toBe(images[2].url)
        expect(getByTestId('image-test').alt).toBe(images[2].description)
    })
    test('previous button click event test', () => {
        const {getByTestId} = render(<App/>)
        const btnNext = getByTestId('next')
        const btnPrevious = getByTestId('previous')
        expect(getByTestId('image-test').src).toBe(images[0].url)
        expect(getByTestId('image-test').alt).toBe(images[0].description)
        fireEvent.click(btnNext)
        expect(getByTestId('image-test').src).toBe(images[1].url)
        expect(getByTestId('image-test').alt).toBe(images[1].description)
        fireEvent.click(btnPrevious)
        expect(getByTestId('image-test').src).toBe(images[0].url)
        expect(getByTestId('image-test').alt).toBe(images[0].description)
    })
})

describe("Button click validation for first and last image", () => {
    test('previous button click event test when in first image', () => {
        const {getByTestId} = render(<App/>)
        const btnPrevious = getByTestId('previous')
        fireEvent.click(btnPrevious)
        expect(getByTestId('image-test').src).toBe(images[images.length-1].url)
        expect(getByTestId('image-test').alt).toBe(images[images.length-1].description)
    })

    test('next button click event test when in last image', () => {
        const {getByTestId} = render(<App/>)
        const btnNext = getByTestId('next')
        for(let i = 0 ; i <= images.length-1; i++){
            fireEvent.click(btnNext)
        } 
        expect(getByTestId('image-test').src).toBe(images[0].url)
        expect(getByTestId('image-test').alt).toBe(images[0].description)
    })
})