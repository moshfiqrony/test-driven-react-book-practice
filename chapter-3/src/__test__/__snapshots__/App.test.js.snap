// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App's architecture test are the Figure -> image and figcaption rendered correctly? 1`] = `
<figure>
  <img
    alt="AI Image Enlarger - Image Enlarger Online Tool Free"
    data-testid="image-test"
    src="https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQYmn-AewAIxk9ATJC3VOb26sVuSDkdqzDNieX9pov466JFno1k&usqp=CAU"
  />
  <figcaption
    data-testid="figcaption-test"
  >
    <strong>
      AI Image Enlarger - Image Enlarger Online Tool Free
    </strong>
  </figcaption>
</figure>
`;
